Bluetooth: Host: Fix incorrect build assert

jhedberg opened the pull request:
We want to make sure `struct closure` fits in the user data, so a user data size of `sizeof(struct closure)` is valid.

jhedberg committed:
Bluetooth: Host: Fix incorrect build assert

We want to make sure `struct closure` fits in the user data, so a user data
size of `sizeof(struct closure)` is valid.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg commented:
This fixes a regression introduced by commit 28535fe2f27a027b8f0712ce724224b250261d6b

jhedberg commented:
Btw, seems like we may need to backport this to 3.7 as well?

jori-nordic commented:
how did it not break CI then?

jhedberg commented:
> This PR seems to be triggering a test case failure:
> 
> ```
> ===================================================================
> START - test_bt_id_set_adv_private_addr_fails
> 
>     Assertion failed at /__w/zephyr/zephyr/tests/bluetooth/host/id/bt_id_set_adv_own_addr/src/test_suite_invalid_inputs.c:111: bt_id_set_adv_own_addr_invalid_inputs_test_bt_id_set_adv_private_addr_fails: (err < 0 is false)
> Unexpected error code '0' was returned
>  at test function
>  FAIL - test_bt_id_set_adv_private_addr_fails in 0.000 seconds
> ===================================================================
> ```
> 
> Not sure if it's related to this PR, or something that was always there.

This looks like a regression from commit 0a8bbbda - the test is expecting `ENOTSUP` whereas the code today returns `-EINVAL`. @jori-nordic should the implementation or the test be fixed?

jhedberg commented:
> This looks like a regression from commit [0a8bbbd](https://github.com/zephyrproject-rtos/zephyr/commit/0a8bbbda4aacae94a249ea49949ed9e546ed09b5) - the test is expecting `-ENOTSUP` whereas the code today returns `-EINVAL`. @jori-nordic should the implementation or the test be fixed?

Sorry, this was not it (I was looking at the wrong branch in `bt_id_set_adv_own_addr()` and the test output clearly states a return value of `0`. I'm still wondering what broke this and why it wasn't caught.

sjanc reviewed (dismissed):
I think this worked because currently userdata is set with some extras bytes (closure is 8 on 32bit, but userdata is set to 16,  and same on 64bit 16->32)

config BT_CONN_TX_USER_DATA_SIZE
        int
        default 32 if 64BIT
        default 16

So with defaults it doesn't hit this assert.

jhedberg commented:
For whatever reason it seems that several of these tests have been ignored for quite some time by upstream CI, which has resulted in them regressing. The CI failure in this PR seems to be another example of it (unrelated to this PR).

jhedberg commented:
> This PR seems to be triggering a test case failure:
> 
> ```
> ===================================================================
> START - test_bt_id_set_adv_private_addr_fails
> 
>     Assertion failed at /__w/zephyr/zephyr/tests/bluetooth/host/id/bt_id_set_adv_own_addr/src/test_suite_invalid_inputs.c:111: bt_id_set_adv_own_addr_invalid_inputs_test_bt_id_set_adv_private_addr_fails: (err < 0 is false)
> Unexpected error code '0' was returned
>  at test function
>  FAIL - test_bt_id_set_adv_private_addr_fails in 0.000 seconds
> ===================================================================
> ```
> 
> Not sure if it's related to this PR, or something that was always there.

Simply tracing through the implementation with the inputs given in this test case, I don't see why an error should be returned, i.e. the test case seem wrong to me (perhaps it should be removed?). Any comment from @jori-nordic or @ahmedmoheb-nordic (the original author of this test)?

jori-nordic committed:
tests: Bluetooth: disable optimizations in unit test

Very very weird behavior.

Deleting the `options & BT_LE_ADV_OPT_USE_NRPA` branch in
`id.c:bt_id_set_adv_own_addr` makes the test pass. But add a log in that
branch, and you'll see that we never take it in the UT binary..

I don't have time for this, maybe some compiler guru or C language
lawyer can look at that decompiled function and figure out the root
cause.

Someone should also really fix the Bluetooth unit tests and make them
run on any Bluetooth related PR. Also nightly.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jhedberg commented on tests/bluetooth/host/id/bt_id_set_adv_own_addr/prj.conf:
```diff
@@ -7,3 +7,6 @@ CONFIG_ASSERT=y
 CONFIG_ASSERT_LEVEL=2
 CONFIG_ASSERT_VERBOSE=y
 CONFIG_ASSERT_ON_ERRORS=y
+
+CONFIG_NO_OPTIMIZATIONS=y
+CONFIG_LTO=n
```
I'd at least add a comment above these with a short explanation, so that one doesn't have to go full-on Sherlock Holmes and track down the explanation from the commit that introduced them.

carlescufi replied:
Can @jori-nordic send it as a follow-up? This is blocking all development right now.

jhedberg replied:
> Can @jori-nordic send it as a follow-up? This is blocking all development right now.

That's ok with me, assuming the CI tests pass (they're still running when I write this)

jori-nordic reviewed (approved):
🚑

jori-nordic commented:
for future investigation: seems `CONFIG_LOG_USE_VLA=n` fixes the test 😫
